packages/flutter_tools/lib/src/macos/xcode.dart

```diff
@@ -185,3 +189,303 @@ class Xcode {
+    //    "modelName" : "iPhone 6s",
+    //    "name" : "iPhone",
+    //    "error" : {
+    //      "code" : -9,
+    //      "failureReason" : "",
+    //      "underlyingErrors" : [
+    //        {
+    //          "code" : 5,
+    //          "failureReason" : "allowsSecureServices: 1. isConnected: 0. Platform: <DVTPlatform:0x7f804ce32880:'com.apple.platform.iphoneos':<DVTFilePath:0x7f804ce32800:'\/Users\/magder\/Applications\/Xcode_11-3-1.app\/Contents\/Developer\/Platforms\/iPhoneOS.platform'>>. DTDKDeviceIdentifierIsIDID: 0",
+    //          "description" : "📱<DVTiOSDevice (0x7f801f190450), iPhone, iPhone, 13.3 (17C54), d83d5bc53967baa0ee18626ba87b6254b2ab5418> -- Failed _shouldMakeReadyForDevelopment check even though device is not locked by passcode.",
+    //          "recoverySuggestion" : "",
+    //          "domain" : "com.apple.platform.iphoneos"
+    //        }
+    //      ],
+    //      "description" : "iPhone is not paired with your computer.",
+    //      "recoverySuggestion" : "To use iPhone with Xcode, unlock it and choose to trust this computer when prompted.",
+    //      "domain" : "com.apple.platform.iphoneos"
+    //    }
+    //  },
+    //  {
+    //    "simulator" : false,
+    //    "operatingSystemVersion" : "13.3 (17C54)",
+    //    "interface" : "usb",
+    //    "available" : false,
+    //    "platform" : "com.apple.platform.iphoneos",
+    //    "modelCode" : "iPhone8,1",
+    //    "identifier" : "d83d5bc53967baa0ee18626ba87b6254b2ab5418",
+    //    "architecture" : "arm64",
+    //    "modelName" : "iPhone 6s",
+    //    "name" : "iPhone",
+    //    "error" : {
+    //      "code" : -9,
+    //      "failureReason" : "",
+    //      "description" : "iPhone is not paired with your computer.",
+    //      "domain" : "com.apple.platform.iphoneos"
+    //    }
+    //  }
+    // ...
+
+    if (!deviceProperties.containsKey('error')) {
+      return null;
+    }
+
+    final Map<String, dynamic> error = deviceProperties['error'] as Map<String, dynamic>;
+
+    final StringBuffer errorMessage = StringBuffer('Error: ');
+
+    if (error.containsKey('description')) {
+      final String description = error['description'] as String;
+      errorMessage.write(description);
+      if (!description.endsWith('.')) {
+        errorMessage.write('.');
+      }
+    } else {
+      errorMessage.write('Xcode pairing error.');
+    }
+
+    if (error.containsKey('recoverySuggestion')) {
+      final String recoverySuggestion = error['recoverySuggestion'] as String;
+      errorMessage.write(' $recoverySuggestion');
+    }
+
+    if (error.containsKey('code') && error['code'] is int) {
+      final int code = error['code'] as int;
+      errorMessage.write(' (code $code)');
+    }
+
+    return errorMessage.toString();
+  }
+
+  /// List of all devices reporting errors.
+  Future<List<String>> getDiagnostics() async {
+    final List<dynamic> allAvailableDevices = await _getAllDevices();
```

Comment:
It looks like this is called shortly after an empty return from `getAllConnectedDevices()`. Would it make sense to cache the result of `_getAllDevices()` to avoid shelling out to `xcdevice` again?